bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java

```diff
@@ -511,4 +511,69 @@ public State getStateAs(Set<Item> items, Class<? extends State> stateClass) {
         }
     }
 
+    /**
+     * This does a threshold operation. If at least one item is over 'upperLimit',
+     * the 'upperState' is returned. Only if all items are under 'lowerLimit',
+     * the 'lowerState' is returned, otherwise nothing (null) is returned.
+     */
+    static class Threshold implements GroupFunction {
+
+        protected final State activeState;
+        protected final State passiveState;
+        protected final DecimalType upperLimit;
+        protected final DecimalType lowerLimit;
+
+        public Threshold(State activeValue, State passiveValue, DecimalType upperLimit, DecimalType lowerLimit) {
+            if (activeValue == null || passiveValue == null || upperLimit == null || lowerLimit == null) {
+                throw new IllegalArgumentException("Parameters must not be null!");
+            }
+            if (upperLimit.compareTo(lowerLimit) < 0) {
+                throw new IllegalArgumentException("Parameter upperLimit must not be lower then lowerLimit!");
+            }
+            this.activeState = activeValue;
+            this.passiveState = passiveValue;
+            this.upperLimit = upperLimit;
+            this.lowerLimit = lowerLimit;
+        }
+
+        @Override
+        public State calculate(Set<Item> items) {
+
+            if (items != null && items.size() > 0) {
+                // TODO Auto-generated method stub
+                boolean isLower = true;
+                for (Item item : items) {
+                    if (upperLimit.compareTo((DecimalType)item.getStateAs(upperLimit.getClass())) < 0) {

```

Comment:
We could extract the item state into a local variable

    DecimalType itemState = (DecimalType) item.getStateAs(DecimalType.class);

If either of the limits is actually a subclass of `DecimalType` that won’t affect the result of `compareTo`.